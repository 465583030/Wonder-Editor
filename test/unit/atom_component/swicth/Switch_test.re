open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

let _ =
  describe("Swicth", () => {
    let sandbox = getSandboxDefaultVal();

    let buildSwitch = isOpen =>
      ReactTestRenderer.create(
        <Switch
          openText="run"
          openFunc=(() => ())
          closeText="stop"
          closeFunc=(() => ())
          isOpen
        />,
      );

    beforeEach(() => {
      sandbox := createSandbox();
      WonderLog.Wonder_Console.makeObjInToWindow();
    });
    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));

    describe("test workflow", () => {
      describe("test initial:is close", () => {
        test("show openText", () =>
          buildSwitch(false) |> ReactTestTool.createSnapshotAndMatch
        );
        test("click button should execute onOpen func", () => {
          open Switch;

          let state = {isOpen: false};

          let onCloseFunc = createEmptyStubWithJsObjSandbox(sandbox);
          let onOpenFunc = createEmptyStubWithJsObjSandbox(sandbox);

          let state =
            Switch.reducer(onOpenFunc, onCloseFunc, ChangeState, state)
            |> ReactTool.getUpdateState;

          (onOpenFunc |> getCallCount, state.isOpen) |> expect == (1, true);
        });
      });

      describe("test initial:is open", () => {
        test("show openText", () =>
          buildSwitch(true) |> ReactTestTool.createSnapshotAndMatch
        );
        test("click button should execute onClose func", () => {
          open Switch;

          let state = {isOpen: true};

          let onCloseFunc = createEmptyStubWithJsObjSandbox(sandbox);
          let onOpenFunc = createEmptyStubWithJsObjSandbox(sandbox);

          let state =
            Switch.reducer(onOpenFunc, onCloseFunc, ChangeState, state)
            |> ReactTool.getUpdateState;

          (onCloseFunc |> getCallCount, state.isOpen) |> expect == (1, false);
        });
      });
    });
  });