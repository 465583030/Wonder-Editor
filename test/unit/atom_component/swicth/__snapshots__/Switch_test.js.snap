// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swicth test workflow click button to open show closeText 1`] = `
<article
  className="wonder-switch"
>
  <button
    onClick={[Function]}
  >
    stop
  </button>
</article>
`;

exports[`Swicth test workflow test initial:is close show openText 1`] = `
<article
  className="wonder-switch"
>
  <button
    onClick={[Function]}
  >
    run
  </button>
</article>
`;
